File: src/ollama_hud/services/chat_log_service.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from pathlib import Path

from ollama_hud.services.ollama_service import (
    ChatMemory,
    build_message_preview,
    select_prompt_memories,
)
from ollama_hud.services.settings_service import CHAT_LOG_PATH, HudSettings


@dataclass(frozen=True)
class ChatLogEntry:
    capture_id: str
    question: str
    memories: tuple[ChatMemory, ...]
    answer: str | None = None
    error: str | None = None
    retry: str = "none"
    thinking: str | None = None


class ChatLogger:
    def __init__(self, path: str | Path = CHAT_LOG_PATH):
        self.path = Path(path)

    def write(self, entry: ChatLogEntry, settings: HudSettings) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        with self.path.open("a", encoding="utf-8") as handle:
            handle.write(_format_entry(entry, settings))


def _format_entry(entry: ChatLogEntry, settings: HudSettings) -> str:
    sent_memories = select_prompt_memories(entry.memories, settings.memory_qa_pairs)
    lines = [
        "=" * 80,
        f"Timestamp: {datetime.now().astimezone().isoformat(timespec='seconds')}",
        f"Model: {settings.model}",
        f"Host: {settings.host}",
        f"Capture ID: {entry.capture_id}",
        "Screenshot attached: yes; payload omitted from this text log",
        f"Screenshot max edge: {settings.screenshot_max_edge}",
        f"Q/A memory configured: {settings.memory_qa_pairs}",
        f"Q/A memory available: {len(entry.memories)}",
        f"Q/A memory sent: {len(sent_memories)}",
        f"Think: {'yes' if settings.think else 'no'}",
        f"Retry: {entry.retry}",
        "",
        "Question:",
        entry.question,
        "",
        "Included Memory:",
    ]
    if sent_memories:
        for index, item in enumerate(sent_memories, start=1):
            lines.extend(
                [
                    f"{index}. Q: {_one_line(item.question)}",
                    f"   A: {_one_line(item.answer)}",
                ]
            )
    else:
        lines.append("(none)")

    lines.extend(["", "Answer:"])
    lines.append(entry.answer if entry.answer is not None else "(none)")
    lines.extend(["", "Thinking:"])
    lines.append(entry.thinking if entry.thinking is not None else "(none)")
    if entry.error is not None:
        lines.extend(["", "Error:", entry.error])
    lines.extend(
        [
            "",
            "Message Preview Sent:",
            build_message_preview(
                entry.question,
                instruction=settings.instruction,
                memories=sent_memories,
                max_memories=settings.memory_qa_pairs,
            ),
        ]
    )
    lines.append("")
    return "\n".join(lines) + "\n"


def _one_line(text: str) -> str:
    return " ".join(part.strip() for part in text.splitlines() if part.strip()).strip()

File: native/src/ChatLogService.cpp
#include "ChatLogService.h"

#include <QDateTime>
#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QTextStream>

namespace {
QString oneLine(const QString &text)
{
    QStringList parts;
    for (const QString &part : text.split('\n')) {
        const QString trimmed = part.trimmed();
        if (!trimmed.isEmpty()) {
            parts.append(trimmed);
        }
    }
    return parts.join(' ').trimmed();
}

QString countText(qint64 count)
{
    return count >= 0 ? QString::number(count) : QStringLiteral("(not returned)");
}

QString durationText(qint64 nanoseconds)
{
    if (nanoseconds < 0) {
        return QStringLiteral("(not returned)");
    }
    return QString::number(static_cast<double>(nanoseconds) / 1000000000.0, 'f', 3) + " s";
}
}

void ChatLogService::write(const ChatLogEntry &entry, const HudSettings &settings, const QString &path)
{
    const QString selectedPath = path.isEmpty() ? SettingsStore::chatLogPath() : path;
    QDir().mkpath(QFileInfo(selectedPath).absolutePath());
    QFile file(selectedPath);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Append)) {
        return;
    }

    const QList<ChatMemory> sentMemories = OllamaService::selectPromptMemories(entry.memories, settings.memoryQaPairs);
    QTextStream out(&file);
    out.setEncoding(QStringConverter::Utf8);
    out << QString(80, '=') << "\n";
    out << "Timestamp: " << QDateTime::currentDateTime().toString(Qt::ISODate) << "\n";
    out << "Model: " << settings.model << "\n";
    out << "Host: " << settings.host << "\n";
    out << "Capture ID: " << entry.captureId << "\n";
    out << "Screenshot attached: yes; payload omitted from this text log\n";
    out << "Screenshot max edge: " << settings.screenshotMaxEdge << "\n";
    out << "Q/A memory configured: " << settings.memoryQaPairs << "\n";
    out << "Q/A memory available: " << entry.memories.size() << "\n";
    out << "Q/A memory sent: " << sentMemories.size() << "\n";
    out << "Think: " << (settings.think ? "yes" : "no") << "\n";
    out << "Retry: " << entry.retry << "\n\n";
    out << "Ollama completion:\n";
    out << "Done reason: " << (entry.doneReason.isEmpty() ? "(not returned)" : entry.doneReason) << "\n";
    out << "Prompt tokens: " << countText(entry.promptEvalCount) << "\n";
    out << "Generated tokens: " << countText(entry.evalCount) << "\n";
    out << "Total duration: " << durationText(entry.totalDurationNs) << "\n";
    out << "Load duration: " << durationText(entry.loadDurationNs) << "\n";
    out << "Prompt evaluation duration: " << durationText(entry.promptEvalDurationNs) << "\n";
    out << "Generation duration: " << durationText(entry.evalDurationNs) << "\n\n";
    out << "Question:\n" << entry.question << "\n\n";
    out << "Included Memory:\n";
    if (sentMemories.isEmpty()) {
        out << "(none)\n";
    } else {
        for (int i = 0; i < sentMemories.size(); ++i) {
            out << i + 1 << ". Q: " << oneLine(sentMemories.at(i).question) << "\n";
            out << "   A: " << oneLine(sentMemories.at(i).answer) << "\n";
        }
    }
    out << "\nAnswer:\n" << (entry.answer.isEmpty() ? "(none)" : entry.answer) << "\n\n";
    out << "Thinking:\n" << (entry.thinking.isEmpty() ? "(none)" : entry.thinking) << "\n";
    if (!entry.error.isEmpty()) {
        out << "\nError:\n" << entry.error << "\n";
    }
    out << "\nMessage Preview Sent:\n";
    out << OllamaService::buildMessagePreview(entry.question, settings.instruction, settings.screenshotContext, sentMemories, settings.memoryQaPairs) << "\n\n";
}
